rfd/00xx-kubernetes-access-cluster-discovery.md

```diff
@@ -0,0 +1,428 @@
+
+```yaml
+apiVersion: v1
+data:
+  mapRoles: |
+    - groups:
+      - system:teleport
+      rolearn: arn:aws:iam::222222222222:role/teleport-role
+      username: system:teleport
+...
+```
+
+This means that this mapping has to exist for each cluster that the operator wants to be discovered and to do that, the operator has to change the `configmap/aws-auth` manually. `eksctl` has a simpler method for appending identities into the `configmap/aws-auth`, but it still requires executing it for every cluster that is expected to be enrolled by Teleport.
+
+```bash
+$  eksctl create iamidentitymapping --cluster  <clusterName> --region=<region> \
+      --arn arn:aws:iam::222222222222:role/teleport-role --group system:teleport --username system:teleport
+```
+
+When a user creates an Amazon EKS cluster, the AWS Identity and Access Management (IAM) entity user or role, such as a federated user that creates the cluster, is automatically granted `system:masters` permissions in the cluster's role-based access control (RBAC) configuration. If Teleport can use the same role as the user that created the cluster, then it is not required to change the `configmap`, but this solution requires that every cluster is created by users with the same IAM Role and has several security implications.
+
+### Cluster enroll process
+
+Teleport supports loading clusters from Kubernetes Kubeconfig during the startup procedure and the file has 3 main sections, `clusters`, `contexts` and `users`:
+
+```yaml
+
+apiVersion: v1
+clusters:
+- cluster:
+    certificate-authority: ca.crt
+    server: https://127.0.0.1:52181
+  name: minikube
+contexts:
+- context:
+    cluster: minikube
+    namespace: default
+    user: minikube
+  name: minikube
+kind: Config
+users:
+- name: minikube
+  user:
+    client-certificate: client.crt
+    client-key: client.key
+```
+
+- `clusters`: defines the `cluster` API and certificate authority.
+- `users`: describes `user` credentials.
+- `contexts`: links a `cluster` and `user` definition.
+
+In Teleport, the `cluster` details are used to create the cluster config and the `users` credentials are stored in credentials map. This map links the cluster name into the credentials. The discovery mechanism will leverage this behavior and dynamically enroll clusters using the same philosophy but without the necessity of writing or editing the kubeconfig file.
+
+For each cluster to enroll, due to limitations described in the [EKS Discovery and IAM](#aws-eks-discovery-and-iam) section, Teleport cannot create access to itself, so the operator has to create the following resources for each cluster he pretends that Teleport enrolls.
+
+The first resource to be created is the `ClusterRole` that grants minimum RBAC permissions to get pods and impersonate groups, users and service accounts. It is possible to run map Teleport role into `system:masters` but for security reasons that is discouraged.
```

Comment:
If Teleport already has access to the cluster (e.g. when running under cluster creator IAM identity, or if it's already present in aws-auth), can it create these resources automatically?

Reply:
it can but since it has `system:masters` permissions we cannot add the mapping into `configmap/aws-auth` because it might cause some troubles and we might remove everyone's access  by overwrite the creators role with another role that is not `system:masters`